File: src/exposureModel.cpp
#include <Rcpp.h>
#include <exposureModel.hpp>

using namespace Rcpp;

exposureModel::exposureModel(SEXP _X, SEXP _ntpt, SEXP _nloc, SEXP _priorMean, SEXP _prec)
{
    int i,j;
    Rcpp::NumericMatrix inX(_X);
    Rcpp::NumericVector inTpt(_ntpt);
    Rcpp::NumericVector inLoc(_nloc);
    Rcpp::NumericVector inPrecision(_prec);
    Rcpp::NumericVector priorMeans(_priorMean);

    nTpt = inTpt(0);
    nLoc = inLoc(0);

    X = Eigen::MatrixXd(inX.nrow(), inX.ncol());
    offset = Eigen::VectorXd(nTpt);
    betaPriorMean = Eigen::VectorXd(priorMeans.size());
    betaPriorPrecision = Eigen::VectorXd(inPrecision.size());


    if (inX.nrow() != ((nTpt) *(nLoc)))
    {
        Rcpp::Rcout << "Error: Covariate matrix has invalid dimensions." 
                    << " X should be (n*t) by p\n";
        Rcpp::stop("error building exposure model");
    }
    if (inPrecision.length() != priorMeans.length() || priorMeans.length() != X.cols())
    {
        Rcpp::Rcout << "Initial parameters have different length then number of columns of X\n";
        Rcpp::Rcout << "ncol(X): " << X.cols() << "\n";
        Rcpp::Rcout << "length(betaPriorMean): " << priorMeans.length() << "\n";
        Rcpp::Rcout << "length(betaPriorPrecision): " << inPrecision.length() << "\n";
        Rcpp::stop("error building exposure model");
    }
    for (i = 0; i < nTpt; i++)
    {
        offset(i) = 1.0;
    }
    for (i = 0; i < X.rows(); i++)
    {
        for (j = 0; j < X.cols(); j++)
        {
            X(i,j) = inX(i,j);
        }
    }
    for (i = 0; i < X.cols(); i++)
    {
        betaPriorMean(i) = priorMeans(i);
        betaPriorPrecision(i) = inPrecision(i);
    }
}

exposureModel::exposureModel(exposureModel* tocopy)
{
    nTpt = (tocopy -> nTpt);
    nLoc = (tocopy -> nLoc);

    Eigen::MatrixXd Xc = (tocopy -> X);
    X = Xc;

    Eigen::VectorXd betaPriorPrecisionc = (tocopy -> betaPriorPrecision);
    betaPriorPrecision = betaPriorPrecisionc;

    Eigen::VectorXd betaPriorMeanc = (tocopy -> betaPriorMean);
    betaPriorMean = betaPriorMeanc;

    Eigen::VectorXd offsetc = (tocopy -> offset);
    offset = offsetc; 
}

int exposureModel::getModelComponentType()
{
    return(LSS_EXPOSURE_MODEL_TYPE);
}

void exposureModel::summary()
{
    Rcpp::Rcout << "Exposure Model Summary\n" << 
                   "----------------------" << 
                   "    dim(X): (" << X.rows() << ", " <<
                   X.cols() << ")\n" << 
                   "    beta prior means: (";
    int sz = betaPriorMean.size();
    int i;
    for (i = 0; i < sz; i++)
    {
        if (i != 0) Rcpp::Rcout << "                       ";
        Rcpp::Rcout << betaPriorMean(i);
        if (i+1 < sz)
        {
            Rcpp::Rcout << ",\n";
        }
        else
        {
            Rcpp::Rcout << ")\n";
        }
    }
    Rcpp::Rcout << "    beta prior precision: (";
    for (i = 0; i < sz; i++)
    {
        if (i != 0) Rcpp::Rcout << "                           ";
        Rcpp::Rcout << betaPriorPrecision(i);
        if (i+1 < sz)
        {
            Rcpp::Rcout << ",\n";
        }
        else
        {
            Rcpp::Rcout << ")\n";
        }
    }
    Rcpp::Rcout << "\n";

                   
}

void exposureModel::setOffset(NumericVector offsets)
{
    if (offsets.length() != (nTpt))
    {
        Rcpp::Rcout << "Error: offsets must have length equal to the number of time points.\n";
        Rcpp::stop("Invalid offsets.");
    }
    int i;
    for (i = 0; i < offsets.length(); i++)
    {
        offset(i) = offsets(i);
    }
}

Rcpp::NumericVector exposureModel::getOffset()
{
    Rcpp::NumericVector out(offset.size());
    int i;
    for (i = 0; i < offset.size(); i++)
    {
        out(i) = offset(i);
    }
    return(out);
}

exposureModel::~exposureModel()
{
    if (prot !=0 ){
        Rcpp::stop("can't delete exposureModel, still being used.\n");
    }
}

RCPP_MODULE(mod_exposureModel)
{
    using namespace Rcpp;
    class_<exposureModel>( "exposureModel" )
    .constructor<SEXP,SEXP,SEXP,SEXP,SEXP>()
    .property("offsets", &exposureModel::getOffset, &exposureModel::setOffset);
}

File: R/exposureModel.R
#' Create an ExposureModel object, describing an epidemic intensity process
#'
#' @param X an $(n*T)$ by $p$ design matrix, where $n$ is the number
#' of locations, $T$ is the number of time points, and $p$ is the number
#' of exposure process parameters. Each column corresponds to a parameter,
#' while each block of $T$ rows corresponds to the time series of covariate
#' values associated with a location. 
#' @param nTpt the number of time points, $T$
#' @param nLoc the number of locations, $N$
#' @param betaPriorPrecision the prior precisions of the $p$ exposure 
#' process parameters
#' @param betaPriorMean the prior means of the $p$ exposure process
#' parameters
#' @param offset a vector of $T$ temporal offset terms, capturing the 
#' relative ammount of aggregated continuous time corresponding to each
#' recorded discrete time point. 
#'
#' @details
#' The exposure process allows the inclusion of both spatially and temporally
#' varying covariates, as well as invariant quantities (intercepts, demographic 
#' features etc.).  
#' @examples exposureModel <- ExposureModel(cbind(1, (1:25)/25),
#'                                          nTpt = 25, nLoc = 1,
#'                                          betaPriorPrecision = 0.1,
#'                                          betaPriorMean = 0)
#' @export
ExposureModel = function(X,nTpt, nLoc,betaPriorPrecision=NA,
                              betaPriorMean=NA,offset=NA)
{
    nBeta = ncol(X)
    checkArgument("X", mustHaveClass("matrix"))

    if (nrow(X) != nTpt * nLoc){
        stop("Invalid data dimensions: X should be a matrix composed of nLoc row-wise blocks of dimension nTpt*p.") 
    } 
    if (length(betaPriorPrecision) == 1 && is.na(betaPriorPrecision))
    {
        print("No prior precision specified, using 0.1")
        betaPriorPrecision = rep(0.1, nBeta)
    }
    else if (length(betaPriorPrecision) == 1)
    {
        betaPriorPrecision = rep(betaPriorPrecision, nBeta)
    }
    if (length(betaPriorMean) == 1 && is.na(betaPriorMean))
    {
        print("No prior mean specified, using zero.")
        betaPriorMean = rep(0, nBeta)
    }
    else if (length(betaPriorMean) == 1)
    {
        betaPriorMean = rep(betaPriorMean, nBeta)
    }
    if (length(offset) == 1 && is.na(offset))
    {
        print("Assuming equally spaced count data.") 
    }
    else if (length(offset) == 1){
        offset = rep(offset, nTpt)
    }
    else if (length(offset) != nTpt){
        stop("Offset must be the of length nTpt")
    }
    if (length(offset) != 1 && max(abs(offset - as.integer(offset))) > 0)
    {
        stop("Offsets must be integers - we need a mapping onto discrete time units.")
    }
    structure(list(X=X,
                   nTpt=nTpt,
                   nLoc=nLoc,
                   betaPriorMean=betaPriorMean,
                   betaPriorPrecision=betaPriorPrecision,
                   offset=offset), class="ExposureModel")
}


